providers/oidc.go

```diff
@@ -186,23 +186,38 @@ func (p *OIDCProvider) redeemRefreshToken(ctx context.Context, s *sessions.Sessi
 		return fmt.Errorf("unable create new session state from response: %v", err)
 	}
 
+	replaceSession(s, newSession)
+	return nil
+}
+
+func replaceSession(s *sessions.SessionState, newSession *sessions.SessionState) {
 	// It's possible that if the refresh token isn't in the token response the
 	// session will not contain an id token.
 	// If it doesn't it's probably better to retain the old one
 	if newSession.IDToken != "" {
 		s.IDToken = newSession.IDToken
+	}
+
+	// Only copy over fields if they are present. Otherwise they might've
+	// originally been set via a ProfileURL call in `EnrichSession` but
+	// are empty in IDToken claims.
+	if newSession.Email != "" {
 		s.Email = newSession.Email
+	}
+	if newSession.User != "" {
 		s.User = newSession.User
+	}
+	if newSession.Groups != nil {
```

Comment:
This could result in old groups (that were actually set via `groups` claim) that the user was completely removed from, resulting in empty `groups` claim in the refreshed IDToken 🤔 -- this would let that membership linger.

Absent another call to the ProfileURL on RefreshSession to really ensure all the proper fields (no way!) -- we have to make a tradeoff based on assumptions I think

Reply:
Why don't we want to use the profile URL on a refresh call? I think that's a reasonable thing to do to make sure we refresh all the information no?

Reply:
I thought we wanted to avoid multiple network round trip calls during a refresh session if at all possible (shave off as many ms of interruption to normal session flow as possible)?

I feel like for OIDC, the `userinfo` fallback is only applicable to <1% of providers, so it feels weird to burden the 99% who properly put claims in the IDToken with this performance penalty.

I'm beginning to think as I dig more into Refresh/Validate session the implications of those calls and how they replace existing session fields throughout our providers have some rough edges in the design 😅

Reply:
Other option is conditional call to `userinfo` / `EnrichSession` if any of these fields are empty in the refreshed session?

But that might lead to architecture complications with other providers with the future design where they embed OIDCProvider and override a few implementations (usually `EnrichSession`).